Write phase 3 articles (#8)
CoherentLabs/StarterGuide: src/content/docs/phase-3-layout-assets-and-styling/laying-out-the-screen.mdx

description: Learn how to lay out your game UI using Gameface's powerful layout system. This section covers best practices for structuring your HTML, using CSS for responsive and scalable layouts, and managing assets like images and fonts to create visually stunning interfaces that perform well across different devices and aspect ratios.
sidebar:
    order: 1

> draft: true

Reviewer: Can we add min and max heading level here so all the headings are properly shown inside the table of contents?

- anchoring something to UI element
- anchoring something to 3D object
- centering on screen

> - centering within control

Reviewer: (Suggestion) Maybe we can wrap the plans of the guides with `<Internal>` so we don't need to iterate after that to delete them

In a standard browser, you have multiple layout algorithms: `block`, `inline`, `grid`, `flex`, and `table`. In Gameface, the layout engine relies exclusively on the <Highlight>Flexbox model</Highlight>. 
Even if you use alternative display types like `display: block`, Gameface simulates them under the hood using `flex` properties. 

Because everything is fundamentally flex-based, every time an element's size, position, or content changes, the engine evaluates how that change affects the surrounding document. 
This calculation process is called a <Highlight>layout pass</Highlight>. Because flexbox distributes space dynamically, elements are highly dependent on their siblings and parents. If one box changes its width, the engine calculates if the boxes next to it need to shrink, stretch, or wrap to a new line.

Gameface's internal layout engine is highly optimized and incredibly fast. However, the environment it runs in is extremely demanding. If a game targets 60 frames per second, the entire game engine has exactly <Highlight>16.6 milliseconds</Highlight> to render a single frame. The UI is usually allocated only a tiny fraction of that time (often 1-2ms). 

> In a standard web browser, a 15ms layout recalculation goes completely unnoticed. In a game engine, it blows your entire frame budget. To ensure your UI runs at peak performance and leaves plenty of CPU room for the game itself, we need to architect our layouts following a few core rules.

Reviewer: Maybe? :D Using `architect` is fine but for me is kind of weird when reading the text. You can also use `build`, `develop` or other word.

```suggestion
In a standard web browser, a 15ms layout recalculation goes completely unnoticed. In a game engine, it blows your entire frame budget. To ensure your UI runs at peak performance and leaves plenty of CPU room for the game itself, we need to design our layouts following a few core rules.
```

## High-Performance Layout Rules

> Before we look at the screen structure, here are the practical "Do's and Don'ts" for styling your elements in Gameface.

Reviewer: ```suggestion
Before we look at the screen structure, here are the practical "Good" and "Bad" practices for styling your elements in Gameface.
```

> ### Define explicit dimensions

Reviewer: For the next headings you can add badges for "Good" and "Bad" or add X and checkmark emojis so the reader knows which is bad and which is good before starting to read the section.

For example:

```suggestion
### [Good badge/checkmark icon] Define explicit dimensions
```

or

```suggestion
### Define explicit dimensions [Good badge/checkmark icon]
```

When an element's size is left `undefined` (or set to `auto`), the engine does not know how big it is. 
It has to pause, dive into the element's children, measure all of their text and content, bubble that size back up, and *then* calculate how to fit it into the flex container. 

:::tip[The explicit size rule]
Always set a `flex-basis`, `width`, or `height` on your elements whenever possible. 
By explicitly defining the boundaries, you tell the engine exactly how much space to allocate immediately, entirely skipping the heavy measuring phase.
:::

### AVOID `align-items: stretch`
In standard web development, stretching items to fill a container is very common. In Gameface, `align-items: stretch` forces the engine to run multiple layout passes. 
It has to measure the children to find the largest one, and then run another pass to force the remaining children to grow and match that size.

Try to use `align-items: center` or `flex-start` and rely on explicit dimensions instead.

> :::danger[Default property]

Reviewer: IMO this should be marked as `caution` and not `danger`

    transform: translate(-50%, -50%); 
}
.crosshair {
    /* Animating scale for bloom effect */
    transition: transform 0.1s ease-out;
    width: 100%;
    height: 100%;
    transform: scale(1.1);
}
```

:::note[Why It Works?]
When the crosshair scales up or down using transforms, the layout engine ignores it. It simply repaints the larger crosshair independently, preserving your frame budget.
:::

#### Scenario 2: Anchoring to a 3D Object
When creating a nameplate that floats above a 3D character's head, the game will send updated 2D screen coordinates (left/top) to the UI every single frame as the camera moves.

Passing these constant updates to an element inside a flex container would force the engine to recalculate the layout repeatedly. 
Instead, we use `absolute positioning` so the engine can independently paint the nameplate at the new coordinates while bypassing the rest of the layout.

```html title="hud.html"
<body>
    <div class="nameplate">
        Player 1
    </div>
</body>
```

```css title="hud.css"
.nameplate {
    /* Absolute is mandatory here */
    position: absolute; 
    
    /* Explicitly size the container */
    width: 15rem;
    height: 3rem;

>     /* The Left/Top values are injected by JavaScript every frame */

Reviewer: ```suggestion
    /* The Left/Top values are injected/changed by JavaScript or Data Binding every frame */
```

    left: 0;
    top: 0;
}
```

:::note[Why It Works?]
Updating `left` and `top` coordinates on an absolutely positioned element only updates that specific node.
 If this nameplate were part of a flex grid, updating its position 60 times a second would force continuous, heavy layout recalculations.
:::

### Flexbox for Widget Alignment
While absolute positioning is perfect for root widgets, it is highly impractical for laying out dynamic content *inside* a widget (like lists, rows, or grids). This is where you should use `display: flex`.

> #### Scenario 3: Centering within a control

Reviewer: Its kind of weird how the headings are structured. You have h3 for absolute position, then you start scenario 1 and 2 that are h4 and after that you have flexbox (h3) followed with h4 scenario 3, etc. Is there is another approach that we can use about this structure?
